core/src/main/scala/org/apache/spark/crypto/CryptoCodec.scala

```diff
@@ -0,0 +1,144 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.crypto
+
+import java.lang.ClassCastException
+
+import scala.reflect.runtime.universe
+
+import org.apache.spark.crypto.CommonConfigurationKeys.SPARK_SECURITY_CRYPTO_CIPHER_SUITE_DEFAULT
+import org.apache.spark.crypto.CommonConfigurationKeys.SPARK_SECURITY_CRYPTO_CIPHER_SUITE_KEY
+import org.apache.spark.crypto.CommonConfigurationKeys.SPARK_SECURITY_CRYPTO_CODEC_CLASSES_AES_CTR_NOPADDING_KEY
+import org.apache.spark.{SparkConf, Logging}
+
+/**
+ * Crypto codec class, encapsulates encryptor/decryptor pair.
+ */
+abstract case class CryptoCodec() {
+  /**
+   *
+   * @return the CipherSuite for this codec.
+   */
+  def getCipherSuite(): CipherSuite
+
+  /**
+   * This interface is only for Counter (CTR) mode. Generally the Encryptor
+   * or Decryptor calculates the IV and maintain encryption context internally.
+   * For example a {@link javax.crypto.Cipher} will maintain its encryption
+   * context internally when we do encryption/decryption using the
+   * Cipher#update interface.
+   * <p/>
+   * The IV can be calculated by combining the initial IV and the counter with
+   * a lossless operation (concatenation, addition, or XOR).
+   * @see http://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_.28CTR.29
+   * @param initIV
+   * @param counter
+   * @param IV
+   */
+  def calculateIV(initIV: Array[Byte], counter: Long, IV: Array[Byte])
+
+  /**
+   * @return Encryptor the encryptor
+   */
+  def createEncryptor: Encryptor
+
+  /**
+   * @return Decryptor the decryptor
+   */
+  def createDecryptor: Decryptor
+
+  /**
+   * Generate a number of secure, random bytes suitable for cryptographic use.
+   * This method needs to be thread-safe.
+   * @param bytes byte array to populate with random data
+   */
+  def generateSecureRandom(bytes: Array[Byte])
+}
+
+object CryptoCodec extends Logging {
+  def getInstance(conf: SparkConf): CryptoCodec = {
+    var name: String = conf.get(SPARK_SECURITY_CRYPTO_CIPHER_SUITE_KEY,
+      SPARK_SECURITY_CRYPTO_CIPHER_SUITE_DEFAULT)
+    getInstance(conf, CipherSuite.convert(name))
+  }
+
+  def getInstance(conf: SparkConf, cipherSuite: CipherSuite): CryptoCodec = {
+    var klasses: List[String] = getCodecClasses(conf, cipherSuite)
+    var codec: CryptoCodec = null
+    for (klass <- klasses) {
+      try {
+        val m = universe.runtimeMirror(getClass.getClassLoader)
+        var c: CryptoCodec = null
+        if (klass.equals("org.apache.spark.crypto.JceAesCtrCryptoCodec")) {
+          val classCryptoCodec = universe.typeOf[org.apache.spark.crypto.JceAesCtrCryptoCodec]
```

Comment:
Why do you need to use reflection here? Why can't you just instantiate the class directly, since it's part of Spark already?